### PR TITLE
Make changes on docker validation

### PR DESCRIPTION
**Pull request by sima-zhu**
The PR is to validate whether packages are good for the Doc.MS before updating docs.
Currently, we have workaround which add docker id+secret to pipeline existing variable group: Release to Github IO secrets.
Create an issue for long term fix: https://github.com/Azure/azure-sdk-tools/issues/2373

Test pipeline on release pipeline: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1225620&view=results
Test pipeline on unified pipeline: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1225618&view=results
Test on no docker id: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1225295&view=results

### CODE REVIEW

File: eng/common/scripts/Update-DocsMsMetadata.ps1 [hunk 1]
```diff
@@ -165,5 +180,12 @@ function UpdateDocsMsMetadataForPackage($packageInfoJsonLocation) {
 
 foreach ($packageInfo in $PackageInfoJsonLocations) {
   Write-Host "Updating metadata for package: $packageInfo"
+  # Add validation step for daily update and release
+  if ($ValidateDocsMsPackagesFn -and (Test-Path "Function:$ValidateDocsMsPackagesFn")) {
+    &$ValidateDocsMsPackagesFn -PackageInfo $packageInfo 
+    if ($LASTEXITCODE -ne 0) {
+      LogError "The package failed Doc.Ms validation. Please fixed the doc and republish to Doc.Ms."
```
Comment (weshaggard) [inline eng/common/scripts/Update-DocsMsMetadata.ps1:187]:
```suggestion
      LogError "The package failed Doc.Ms validation. Please fix the doc and republish to Doc.Ms."
```

File: eng/common/scripts/Update-DocsMsMetadata.ps1 [hunk 2]
```diff
@@ -165,5 +180,12 @@ function UpdateDocsMsMetadataForPackage($packageInfoJsonLocation) {
 
 foreach ($packageInfo in $PackageInfoJsonLocations) {
   Write-Host "Updating metadata for package: $packageInfo"
+  # Add validation step for daily update and release
+  if ($ValidateDocsMsPackagesFn -and (Test-Path "Function:$ValidateDocsMsPackagesFn")) {
+    &$ValidateDocsMsPackagesFn -PackageInfo $packageInfo
```
Comment (weshaggard) [inline eng/common/scripts/Update-DocsMsMetadata.ps1:185]: We should pass in the PackageSourceOverride here as well. Even if someone doesn't provide that parameter it should be a no-op for those.

File: eng/common/scripts/common.ps1 [hunk 3]
```diff
@@ -45,3 +45,4 @@ $GetDocsMsDevLanguageSpecificPackageInfoFn = "Get-${Language}-DocsMsDevLanguageS
 $GetGithubIoDocIndexFn = "Get-${Language}-GithubIoDocIndex"
 $FindArtifactForApiReviewFn = "Find-${Language}-Artifacts-For-Apireview"
 $TestProxyTrustCertFn = "Import-Dev-Cert-${Language}"
+$ValidateDocsMsPackagesFn = "Validate-${Language}-DocMsPackages"
```
Comment (weshaggard) [inline eng/common/scripts/common.ps1:48]: Is there any further common logic we should pull into the eng/common scripts here, like the calling of the docker image assuming the parameters are the same.

Comment (sima-zhu) [reply eng/common/scripts/common.ps1:48]: Call docker image is only one liner code. besides we use different parameter for different language.
Like Java use group id, Python use extra-index-url etc

File: eng/common/scripts/Update-DocsMsMetadata.ps1 (hunk 2 above)
Comment (sima-zhu) [reply eng/common/scripts/Update-DocsMsMetadata.ps1:185]: Then we have to add check null step here before making the call. My preference is to check null right before the logic.

Comment (weshaggard) [reply eng/common/scripts/Update-DocsMsMetadata.ps1:185]: We don't need to add a null check we just blindly pass it and if any consumers need it, they will declare the parameter and they will check it before using it.

File: eng/common/scripts/common.ps1 (hunk 3 above)
Comment (weshaggard) [reply eng/common/scripts/common.ps1:48]: Ok The call is one line of code but I know you are also handling various exit codes in multiple places. However if they all have different parameters they pass to the docker image then I agree there isn't much we can share.

File: eng/common/scripts/Update-DocsMsMetadata.ps1 (hunk 2 above)
Comment (sima-zhu) [reply eng/common/scripts/Update-DocsMsMetadata.ps1:185]: Do you mean we pass something like:

```
&$ValidateDocsMsPackagesFn -PackageInfo $packageInfo  $PackageSourceOverride DocValidationImageId
```

If we pass something like below, poweshell will complain taking null object
```
&$ValidateDocsMsPackagesFn -PackageInfo $packageInfo  -PackageSourceOverride  $PackageSourceOverride -DocValidationImageId $DocValidationImageId
```

Comment (sima-zhu) [reply eng/common/scripts/Update-DocsMsMetadata.ps1:185]: This script is the only place to call ValidateDocsMsPackagesFn currently.

Comment (weshaggard) [reply eng/common/scripts/Update-DocsMsMetadata.ps1]: I would expect the following to work:
```
&$ValidateDocsMsPackagesFn -PackageInfo $packageInfo  -PackageSourceOverride  $PackageSourceOverride -DocValidationImageId $DocValidationImageId
```
What error are you seeing?

File: eng/common/scripts/Update-DocsMsMetadata.ps1 [hunk 4]
```diff
@@ -176,7 +194,18 @@ function UpdateDocsMsMetadataForPackage($packageInfoJsonLocation) {
...
   Write-Host "Updating metadata for package: $packageInfo"
-  UpdateDocsMsMetadataForPackage $packageInfo
+
+  # Convert package metadata json file to metadata json property.
+  $packageInfo = GetPackageInfoJson $packageInfoLocation
+  # Add validation step for daily update and release
+  if ($ValidateDocsMsPackagesFn -and (Test-Path "Function:$ValidateDocsMsPackagesFn")) {
+    &$ValidateDocsMsPackagesFn -PackageInfo $packageInfo -PackageSourceOverride $PackageSourceOverride -DocValidationImageId $DocValidationImageId
+    if ($LASTEXITCODE -ne 0) {
+      LogError "The package failed Doc.Ms validation. Check https://aka.ms/azsdk/docs/docker for how to reproduce the errors"
```
Comment (weshaggard) [inline eng/common/scripts/Update-DocsMsMetadata.ps1:206]:
```suggestion
      LogError "The package failed Doc.Ms validation. Check https://aka.ms/azsdk/docs/docker for more details on how to diagnose this issue."
```

**Review by weshaggard — APPROVED** [review]

File: eng/common/scripts/Update-DocsMsMetadata.ps1 [hunk 5]
```diff
@@ -176,7 +194,18 @@ function UpdateDocsMsMetadataForPackage($packageInfoJsonLocation) {
...
+
+  # Convert package metadata json file to metadata json property.
+  $packageInfo = GetPackageInfoJson $packageInfoLocation
+  # Add validation step for daily update and release
+  if ($ValidateDocsMsPackagesFn -and (Test-Path "Function:$ValidateDocsMsPackagesFn")) {
+    &$ValidateDocsMsPackagesFn -PackageInfo $packageInfo -PackageSourceOverride $PackageSourceOverride -DocValidationImageId $DocValidationImageId
+    if ($LASTEXITCODE -ne 0) {
+      LogError "The package failed Doc.Ms validation. Check https://aka.ms/azsdk/docs/docker for more details on how to diagnose this issue."
+      exit 1
+    }
```
Comment (benbp) [inline eng/common/scripts/Update-DocsMsMetadata.ps1:208]:
```suggestion
    if ($LASTEXITCODE) {
      LogError "The package failed Doc.Ms validation. Check https://aka.ms/azsdk/docs/docker for more details on how to diagnose this issue."
      exit $LASTEXITCODE
    }
```

**Review by benbp — APPROVED** [review]

**Pull request merged**